plugins/perspective_api/newhelm/annotators/perspective_api.py

```diff
@@ -127,33 +137,73 @@ def _process_analyze_comment_response(self, response: Dict) -> Dict[str, float]:
         return flattened
 
 
-def _batch_execute_requests(client: discovery.Resource, requests: List) -> List:
+def _batch_execute_requests(
+    client: discovery.Resource, requests: List, num_retries: int, rng: random.Random
+) -> List:
     """Wrapper around Google's batch API.
 
     This can give significant speedup. For example for PerspectiveAPI, batching
     25 requests is about 15x faster than doing each as separate calls.
     https://googleapis.github.io/google-api-python-client/docs/batch.html
     """
 
-    batch_request: BatchHttpRequest = client.new_batch_http_request()
+    errors = [None] * len(requests)
     responses: List[Dict] = [{}] * len(requests)
 
     def _callback(request_id: str, response: Dict, error: HttpError):
+        index = int(request_id)
         if error:
-            raise error
-        responses[int(request_id)] = response
-
-    for i, request in enumerate(requests):
-        batch_request.add(
-            request=request,
-            request_id=str(i),
-            callback=_callback,
-        )
-    # Fills up responses using the callback
-    batch_request.execute()
+            errors[index] = error
+        else:
+            # Clear any past errors
+            errors[index] = None
+        responses[index] = response
+
+    # Keep track of what requests have not yet successfully gotten a response
+    needs_call = list(range(len(requests)))
+    for retry_count in range(num_retries + 1):
+        if retry_count > 0:
+            # Perform exponential backoff
+            sleep_amount = rng.random() * 2**retry_count
+            logging.info("Performing exponential backoff. Sleeping:", sleep_amount)
+            time.sleep(sleep_amount)
+
+        # Build up a batch
+        batch_request: BatchHttpRequest = client.new_batch_http_request()
+        for i in needs_call:
+            batch_request.add(
+                request=requests[i],
+                request_id=str(i),
+                callback=_callback,
+            )
+        # Fills up responses using the callback
+        batch_request.execute()
```

Comment:
That sounds like a good thing to check and then document with a test.

Reply:
My apologies, I should have linked directly to the the part of the library where that happens:
https://github.com/googleapis/google-api-python-client/blob/369737102a0e944467b18617b2971db25bf85a6c/googleapiclient/http.py#L1500-L1501

Reply:
I didn't get any 503s on this line even when over 400 requests in the batch were returning 503s. 

> That sounds like a good thing to check and then document with a test.

I'm not sure how to write an automated test for that. To do it realistically, we'd need to intentionally overload PerspectiveAPI, which sounds like setting up a DoS attack. If we fake it, its not really testing if that behavior could happen for real.

Reading the docs, [non-batch execute](https://github.com/googleapis/google-api-python-client/blob/369737102a0e944467b18617b2971db25bf85a6c/googleapiclient/http.py#L893-L894) lists that it can raise exceptions for any non 200 response. However  [batch.execute](https://github.com/googleapis/google-api-python-client/blob/369737102a0e944467b18617b2971db25bf85a6c/googleapiclient/http.py#L1540-L1541) doesn't call that out.

Reply:
Nice. Sounds like we're good for the issues we were seeing. Thanks for the info everyone.

Reply:
Sorry I wasn't clear. I think we should manually figure out what happens when we overload something, and then have an automated test that shows that our code is robust for the problem that actually occurs. In addition to proving the behavior is correct, it also serves as documentation of the cases in mind when the code was written.